### PR TITLE
review governance spec

### PR DESCRIPTION
**Pull request by nd-certora**
my review:
notice that some comments are regarding the pdf verification report

### CODE REVIEW

File: security/certora/specs/Governance.spec [hunk 1]
```diff
@@ -51,6 +51,7 @@ methods {
   function getProposalVotingActivationTime(uint256) external returns (uint40) envfree;
   function getProposalSnapshotBlockHash(uint256) external returns (bytes32) envfree;
   function getProposalCancellationFee(uint256) external returns (uint256) envfree;
+  // ND - what the difference between getProposalCount and  getProposalsCount ?
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:54]: duplicate. fixed

File: security/certora/specs/Governance.spec [hunk 2]
```diff
@@ -135,15 +136,33 @@ rule consecutiveIDs(method f) filtered
...
 // For initialized property, the proposal list is not empty
+/* ND:  since this is env dependent we are not checking:
+  require getProposalState(e, proposalId) != IGovernanceCore.State.Null  
+   => getPayloadLength(proposalId) > 0
+  f(e2,args)
+  assert require getProposalState(e3, proposalId) != IGovernanceCore.State.Null  
+   => getPayloadLength(proposalId) > 0
+
+  Can this be an issue? 
+  usually time dependnet properties should be proven as a rule
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:147]: Converted to rules

File: security/certora/specs/Governance.spec [hunk 3]
```diff
@@ -135,15 +136,33 @@ rule consecutiveIDs(method f) filtered
...
       requireInvariant empty_payloads_if_uninitialized_proposal(proposalId);
     }
   }
 // Same property just referring directly to the storage
+/*
+ND: since getProposalStateVariable is a certora harneess function we need a rule checking 
+
+getProposalStateVariable(proposalId) != getProposalState(e,proposalId)
+
+I see many of our rules are on getProposalState() and some are also on getProposalStateVariable() but when and why?
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:164]: We really need getProposalState() - that's the contract function that retrieves the state. All contract functions use this function.

getProposalStateVariable() is a getter to the storage. It's an extra check.

File: security/certora/specs/Governance.spec [hunk 4]
```diff
@@ -222,7 +246,10 @@ rule proposal_after_voting_portal_invalidate{
   IGovernanceCore.State state_after = getProposalState(e3, proposalId);
 
   assert !voting_portal_approved_after => (state_before != state_after => isTerminalState(state_after));
-
+  /* ND- what about the other direction
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:249]: The 2 properties are equivalent.
P1: !voting_portal_approved_after => (state_before != state_after => isTerminalState(state_after))
P2: (state_before != state_after && !isTerminalState(state_after)) => voting_portal_approved_after
They have the same truth table.

  voting_portal_approved_after | state_before != state_after | isTerminalState(state_after) | P1 | P2 
  ---------------------------------------------------------------------------------------------------
             0                 |              0              |             0                | T  | T 
             0                 |              0              |             1                | T  | T 
             0                 |              1              |             0                | F  | F 
             0                 |              1              |             1                | T  | T 
             1                 |              0              |             0                | T  | T 
             1                 |              0              |             1                | T  | T 
             1                 |              1              |             0                | T  | T 
             1                 |              1              |             1                | T  | T

File: security/certora/specs/Governance.spec [hunk 5]
```diff
@@ -241,13 +268,24 @@ rule insufficient_proposition_power(method f) filtered { f -> !f.isView}{
   mathint creator_power = _GovernancePowerStrategy.getFullPropositionPower(e,getProposalCreator(proposalId));
   mathint voting_config_min_power = getMinPropositionPower(getVotingConfig(getProposalAccessLevel(proposalId))) * PRECISION_DIVIDER(); //uint56
 
+  /* ND - it's a but style but require at end of a rule is a bit strange, I try to keep the precondition, state, postcondition style
+  and long expressions are also hard to follow,
+  so how about? 
+
+  bool belowThreshold = (state1 != state2 && creator_power <= voting_config_min_power );
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:275]: agree

File: security/certora/specs/Governance.spec [hunk 6]
```diff
@@ -570,7 +616,7 @@ uint128 forVotes;
     else if (sel ==3) {queueProposal(e, proposalId, forVotes, againstVotes);}
     else if (sel ==4) {executeProposal(e, proposalId);}
     else if (sel ==5) {cancelProposal(e, proposalId);}
-    else {require false;}
+    else {require false;} // ND - better assert and then filter to only statechaning functions
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:619]: fixed

File: security/certora/specs/Governance.spec [hunk 7]
```diff
@@ -241,13 +268,24 @@ rule insufficient_proposition_power(method f) filtered { f -> !f.isView}{
...
+  */
   require state1 != state2; 
   require creator_power <= voting_config_min_power;
   assert state2 == IGovernanceCore.State.Cancelled || state2 == IGovernanceCore.State.Failed;
 
 }
 
 //pass
+/* ND - is insufficient_proposition_power jsut a subcase of insufficient_proposition_power_allow_time_elapse? 
+ok to have both but maybe share the code in cvl function ? */
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:288]: code sahred

File: security/certora/specs/Governance.spec [hunk 8]
```diff
@@ -266,7 +304,10 @@ rule insufficient_proposition_power_allow_time_elapse(method f) filtered { f ->
   state2 == IGovernanceCore.State.Cancelled || state2 == IGovernanceCore.State.Failed || state2 == IGovernanceCore.State.Expired;
 
 }
+/*
+ND: missign rule that only state_advancing_function are indeed such
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:308]: rule state_changing_function_self_check checks that

File: security/certora/specs/Governance.spec [hunk 9]
```diff
@@ -368,13 +413,13 @@ rule state_cant_decrease{
   assert assert_uint256(state1) <= assert_uint256(state2);
 }
 
-
+// ND - in the pdf we say three expections but there are four listed
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:416]: The PDF lists four conditions:
https://github.com/Certora/aave-governance-v3/blob/check_proposition_power/security/certora/reports/Formal%20Verification%20Report%20Of%20Aave%20Governance%20V3.pdf

File: security/certora/specs/Governance.spec [hunk 10]
```diff
@@ -283,17 +324,19 @@ rule insufficient_proposition_power_time_elapsed_tight_witness(method f) filtere
 
   require state1 != state2; 
   require creator_power <= (voting_config_min_power);
+  /* so you are looking for state2 == IGovernanceCore.State.Expired ? why not just satisfy that ? */
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:327]: done

File: security/certora/specs/Governance.spec [hunk 11]
```diff
@@ -619,6 +665,7 @@ rule immutable_after_creation(method f){
 
 
 // Proposal payloads cannot change.
+/* ND - what about f == createProposal why is this passing */
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:668]: Invariant empty_payloads_iff_uninitialized_proposal ensures that uninitialized proposals have no payloads.

**Comment by nd-certora** [comment]
Thinking about missing specs: I wonder if the following mutations will be caught:
1. in updateRepresentativesForChain move 
`_representatives[msg.sender][chainId] = newRepresentative;` line 186 before:
`    address oldRepresentative = _representatives[msg.sender][chainId];` 
2. only_guardian_can_cancel has many cases so: 
2.1 change   
  ```isVotingPortalApproved(proposal.votingPortal) &&
      proposalCreator != msg.sender &&
      _isPropositionPowerEnough(
        _votingConfigs[proposal.accessLevel],
        _powerStrategy.getFullPropositionPower(proposalCreator)
      ```
      to 
      ```
        isVotingPortalApproved(proposal.votingPortal) ||
      proposalCreator != msg.sender ||
      _isPropositionPowerEnough(
        _votingConfigs[proposal.accessLevel],
        _powerStrategy.getFullPropositionPower(proposalCreator)
      ```
  2.2 remove onlyGuardian from rescueVotingPortal

File: security/certora/specs/Governance.spec (hunk 11 above)
Comment (gadicer) [reply security/certora/specs/Governance.spec:668]: The entire rule passed vacuously and it was removed. Before `createProposal()` the array of payloads is empty `_proposals[proposalId].payloads.length == 0` and we cannot refer to its elements.

File: security/certora/specs/Governance.spec [hunk 12]
```diff
@@ -668,7 +715,8 @@ filtered {f -> !f.isView}
   f(e2, args);
   uint40 voting_activation_time_after = getProposalVotingActivationTime(proposalId);
   bytes32 snapshot_blockhash_after = getProposalSnapshotBlockHash(proposalId);
-
+  /* ND - we are not proving the "can only be set once... " we are proving "can not be changed after activateVoting"
+  either rephrase of add a satisfy rule that shows it can be set */
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:719]: Added `satisfy`.

File: security/certora/specs/Governance.spec [hunk 13]
```diff
@@ -725,7 +773,7 @@ rule proposal_executes_after_cooldown_period(){
 //              From its perspective, execution is sent to a Portal.
 //Property #18: VOTING_TOKENS_CAP in GovernanceCore should be big enough to account for tokens that need to pass multiple slots,
 //               and big enough for at least mig to long term
-
+// ND - what are the above 16 - 18 properties?
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:776]: Property #17 We summarize `CrossChainForwarder.forwardMessage` to NONDET. This function sends payloads to execution chain, and it's called by `executeProposal()`. I added a comment about the summarization to the "assumptions and simplifications" section of the report. So the property seems to hold.
Property #18: VOTING_TOKENS_CAP is not relevant anymore, the constant is not used now. I'll notify the customer.

File: security/certora/specs/Governance.spec [hunk 14]
```diff
@@ -841,12 +890,13 @@ filtered { f -> !state_changing_function(f)}
...
 //
 
-// Property: For any proposal id that wasn't yet created, the cancellation fee must be 0
+// Property 22: For any proposal id that wasn't yet created, the cancellation fee must be 0
 invariant cancellationFeeZeroForFutureProposals(uint256 proposalId) 
     proposalId >= getProposalCount() => getProposalCancellationFee(proposalId) == 0;
 
 // Property: eth balance can cover the total cancellation fee of users
 invariant totalCancellationFeeEqualETHBalance()
+// ND - the report regard fee of live proposals, this is all proposals, no?
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:899]: report is fixed

File: security/certora/specs/Governance.spec [hunk 15]
```diff
@@ -176,20 +196,20 @@ invariant creator_is_not_zero_2(uint256 proposalId)
 invariant empty_payloads_iff_uninitialized_proposal(uint256 proposalId)
       proposalId >= getProposalsCount() <=> getPayloadLength(proposalId) == 0;
 
+/* ND - this is just a subset of the above ? */
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:199]: removed

File: security/certora/specs/Governance.spec [hunk 16]
```diff
@@ -176,20 +196,20 @@ invariant creator_is_not_zero_2(uint256 proposalId)
...
 
-
 invariant null_state_variable_iff_uninitialized_proposal(uint256 proposalId)
       proposalId >= getProposalsCount() <=> getProposalStateVariable(proposalId) == IGovernanceCore.State.Null;
-
+/* ND - subset of null_state_iff_uninitialized_proposal ? */
 invariant null_state_if_uninitialized_proposal(env e, uint256 proposalId)
       proposalId >= getProposalsCount() => getProposalState(e, proposalId) == IGovernanceCore.State.Null;
-
+/* ND - subset of null_state_variable_iff_uninitialized_proposal */
```
Comment (gadicer) [inline security/certora/specs/Governance.spec:212]: removed
